File: src/cpp/config_store.cpp
#include "util.h"
#include <fastly/config_store.h>
#include <fastly/sdk-sys.h>

namespace fastly::config_store {

fastly::expected<ConfigStore> ConfigStore::open(std::string_view name) {
  fastly::sys::config_store::ConfigStore *out;
  fastly::sys::error::FastlyError *err;
  fastly::sys::config_store::m_static_config_store_config_store_open(
      static_cast<std::string>(name), out, err);
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return FSLY_BOX(config_store, ConfigStore, out);
  }
}

fastly::expected<std::optional<std::string>>
ConfigStore::get(std::string_view key) {
  std::string out;
  fastly::sys::error::FastlyError *err;
  auto some{this->cs->get(static_cast<std::string>(key), out, err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>(std::move(out));
  }
}

fastly::expected<bool> ConfigStore::contains(std::string_view key) {
  fastly::sys::error::FastlyError *err;
  bool out{this->cs->contains(static_cast<std::string>(key), err)};
  if (err != nullptr) {
    return fastly::unexpected(err);
  } else {
    return fastly::expected<bool>(out);
  }
}

} // namespace fastly::config_store

File: src/config_store.rs
use std::pin::Pin;

use cxx::CxxString;

use crate::{error::ErrPtr, try_fe};

pub struct ConfigStore(pub(crate) fastly::ConfigStore);

pub fn m_static_config_store_config_store_open(
    name: &CxxString,
    mut out: Pin<&mut *mut ConfigStore>,
    mut err: ErrPtr,
) {
    out.set(Box::into_raw(Box::new(ConfigStore(try_fe!(
        err,
        fastly::ConfigStore::try_open(try_fe!(err, name.to_str()))
    )))))
}

impl ConfigStore {
    pub fn get(&self, key: &CxxString, out: Pin<&mut CxxString>, mut err: ErrPtr) -> bool {
        try_fe!(err, self.0.try_get(try_fe!(err, key.to_str())))
            .map(|val| out.push_str(val.as_ref()))
            .is_some()
    }

    pub fn contains(&self, key: &CxxString, mut err: ErrPtr) -> bool {
        self.0.contains(try_fe!(err, key.to_str()))
    }
}

pub fn f_config_store_config_store_force_symbols(x: Box<ConfigStore>) -> Box<ConfigStore> {
    x
}
